h/static/scripts/directive/annotation.js

```diff
@@ -0,0 +1,730 @@
+      changes.text = draft.text;
+    }
+    if (draft.tags) {
+      changes.tags = draft.tags;
+    }
+    if (draft.permissions) {
+      changes.permissions = draft.permissions;
+    }
+    drafts.update(model, changes);
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#save
+    * @description Saves any edits and returns to the viewer.
+    */
+  vm.save = function() {
+    var newTags;
+    var onFulfilled;
+    var onRejected;
+    var updatedModel;
+
+    if (!model.user) {
+      return flash.info('Please sign in to save your annotations.');
+    }
+
+    if (!validate(vm.annotation)) {
+      return flash.info('Please add text or a tag before publishing.');
+    }
+
+    // Update stored tags with the new tags of this annotation.
+    newTags = vm.annotation.tags.filter(function(tag) {
+      var tags = model.tags || [];
+      return (tags.indexOf(tag.text) === -1);
+    });
+    tags.store(newTags);
+
+    switch (vm.action) {
+      case 'create':
+        updateDomainModel(model, vm.annotation);
+        onFulfilled = function() {
+          $rootScope.$emit('annotationCreated', model);
+          vm.view();
+        };
+        onRejected = function(reason) {
+          flash.error(
+            errorMessage(reason), 'Saving annotation failed');
+        };
+        return model.$create().then(onFulfilled, onRejected);
+
+      case 'edit':
+        updatedModel = angular.copy(model);
+        updateDomainModel(updatedModel, vm.annotation);
+        onFulfilled = function() {
+          angular.copy(updatedModel, model);
+          $rootScope.$emit('annotationUpdated', model);
+          vm.view();
+        };
+        onRejected = function(reason) {
+          flash.error(
+            errorMessage(reason), 'Saving annotation failed');
+        };
+        return updatedModel.$update({
+          id: updatedModel.id
+        }).then(onFulfilled, onRejected);
+    }
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#reply
+    * @description
+    * Creates a new message in reply to this annotation.
+    */
+  vm.reply = function() {
+    var id = model.id;
+    var references = model.references || [];
+    var reply;
+    var uri = model.uri;
+
+    if (typeof references === 'string') {
+      references = [references];
+    }
+
+    references = references.concat(id);
+
+    reply = annotationMapper.createAnnotation({
+      references: references,
+      uri: uri
+    });
+    reply.group = model.group;
+
+    if (session.state.userid) {
+      if (permissions.isShared(model.permissions, model.group)) {
+        reply.permissions = permissions.shared(reply.group);
+      } else {
+        reply.permissions = permissions.private();
+      }
+    }
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#render
+    * @description Called to update the view when the model changes.
+    */
+  vm.render = function() {
+    var documentTitle;
```

Comment:
Ah, I was following Douglas Crockford's advice to put all your `var`s at the top of the function in JavaScript, because that's where they're scoped to. Do you want all variables to be defined where they're first used instead, then?